src/docker.c

```diff
@@ -0,0 +1,1754 @@
+              "Proportional System seconds: %lu, CPU seconds: %lu,"
+              "Num active cores: %d.", i,
+              stats[i]->percent_used, proportional_footprint, stats[i]->usage,
+              active_cores);
+        stats[i]->percent_used = 0.00;
+        stats[i]->percent_idle = 100.00;
+      } else {
+        stats[i]->percent_idle = 100.00 - stats[i]->percent_used;
+      }
+    }
+    if (c_avl_insert(cpu_hist_values, (void *) container_id,
+           (void *) old_stats) < 0) {
+      ERROR("docker: c_avl_insert failed!");
+      return DOCKER_ERROR;
+    }
+  }
+  return 0;
+}
+
+/* Purpose: Computes memory utilization percentage for used and free states by
+ *           the container since the last time the Docker API was queried.
+ *
+ * Params: (memory_stats_t *) stats: memory_stats_t struct which stores memory
+ *                                     metrics.
+ * Returns: int
+ *          0 in case of success.
+ *          -1 in case of error.
+ **/
+static int compute_memory_stats(memory_stats_t *stats) {
+  if (stats == NULL) {
+    ERROR("docker: compute_memory_stats. memory stats NULL");
+  }
+  stats->free = stats->limit - stats->usage;
+  if (stats->limit == 0) return DOCKER_ERROR;
+  stats->percent_used = (stats->usage * 100.00)/(stats->limit);
+  stats->percent_free = 100.00 - stats->percent_used;
+  return 0;
+}
+
+//==============================================================================
+//==============================================================================
+//==============================================================================
+// Functionality to parse the Docker Stats JSON response and retrieve stats.
+//==============================================================================
+//==============================================================================
+//==============================================================================
+
+/* Purpose: Retrieve the list of container IDs from the Docker API response.
+ *
+ * Params: (char ***) container_list: Pointer to a list of strings which hold
+ *                                      the container IDs after the function 
+ *                                      returns.
+ *          (char *) response_buffer: Docker API response string.
+ *
+ * Returns: N/A
+ *
+ **/
+static int extract_container_ids_from_response(char ***container_list,
+    char *response_buffer) {
+  yajl_val node;
+  char errbuf[1024];
+  int num_containers = -1;
+  char **list;
+  node = yajl_tree_parse(response_buffer, errbuf, sizeof(errbuf));
+  if (node == NULL) {
+    if (strlen(errbuf)) {
+      ERROR("docker: parse_error: %s.\n", errbuf);
+    } else {
+      ERROR("docker: extract_container_ids_from_response: parse_error.\n");
+    }
+    return num_containers;
+  }
+  if (node->u.array.len == 0) {
+    yajl_tree_free(node);
+    return 0;
+  }
+
+  const char *id_path[] = { "Id", (const char *) 0 };
+  if (YAJL_IS_ARRAY(node)) {
```

Comment:
Do you need to check this? Up on line 640 you just accessed the u.array.len property without checking if it was an array first.